core/primitives/src/time.rs

```diff
@@ -0,0 +1,146 @@
+use std::borrow::Cow;
+use std::collections::HashMap;
+use std::default::Default;
+use std::mem;
+use std::sync::{Arc, Mutex, Once};
+use std::time;
+
+use chrono;
+
+pub use chrono::Utc;
+pub use std::time::Instant;
+
+pub struct UtcProxy;
+pub struct InstantProxy;
+
+pub struct TimeTravelSingleton {
+    pub last_check_utc: chrono::DateTime<chrono::Utc>,
+    pub last_check_instant: time::Instant,
+    pub diff: i64,
+    pub rate: f64,
+    pub proxify: HashMap<FileLocation, bool>,
+}
+
+#[derive(Eq, PartialEq, Hash, Debug)]
+pub struct FileLocation {
+    pub file: Cow<'static, str>,
+    pub line: u32,
+}
+
+#[derive(Clone)]
+struct SingletonReader {
+    // Since we might be used in many threads, we need to protect
+    // concurrent access
+    inner: Arc<Mutex<TimeTravelSingleton>>,
+}
+
+impl Default for TimeTravelSingleton {
+    fn default() -> Self {
+        Self {
+            last_check_utc: chrono::Utc::now(),
+            last_check_instant: time::Instant::now(),
+            diff: 0,
+            rate: 1.0,
+            proxify: HashMap::new(),
+        }
+    }
+}
+
+impl TimeTravelSingleton {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub fn set(value: Self) {
+        SingletonReader::set(value)
+    }
+}
+
+static mut SINGLETON: *const SingletonReader = 0 as *const SingletonReader;
+static ONCE: Once = Once::new();
+
+impl SingletonReader {
+    fn get() -> SingletonReader {
+        unsafe {
+            ONCE.call_once(|| {
+                SingletonReader::set(TimeTravelSingleton::default());
+            });
+
+            (*SINGLETON).clone()
+        }
+    }
+
+    fn set(value: TimeTravelSingleton) {
+        let singleton = SingletonReader { inner: Arc::new(Mutex::new(value)) };
+        unsafe {
+            SINGLETON = mem::transmute(Box::new(singleton));
+        }
+    }
+}
+
```

Comment:
I don't think the unsafe is correct. `TimeTravelSingleton::set` is a public API of the crate, and calling it concurrently results in a race on the `SINGLETON` variable. 

Taking a step back, I don't think this *needs* manually written unsafe. I am a bit confused what's the intent here. It seems like `get` can overwrite a previous set, which feels like a bug.

If you need to initialize something once, once_cell crate will work (it's already among the deps).

If yon needs a gobally swappable singleton, take a look at the `arc_swap` crate.

Reply:
You are right, there's a race. I could address these issues, but I will introduce `arc_swap` in this PR as you suggested because it will eliminate the need for unsafe code.

here's how I'd address the race by changing `SINGLETON`:
```
static mut SINGLETON: AtomicPtr<SingletonReader> = AtomicPtr::new(0 as *const SingletonReader);
```

Reply:
Yup, the new version looks much better :+1: !